### ISSUE TITLE
ApiReponse with type is an Advanced Types

### ISSUE THREAD

**Issue by workfel** [nestjs/swagger#41]
Hi, 

I have an generic class who is return on all my request 
```ts
import { ApiModelProperty } from '@nestjs/swagger';

export class ServerApiResponse<T> {
  
  @ApiModelProperty()
  value: T;
  
  @ApiModelProperty({ type: Number })
  pages: number;


  @ApiModelProperty({ type: Number })
  page: number;

  @ApiModelProperty({ type: Number })
  maxByPage: number;


  @ApiModelProperty({ type: Number })
  count: number;

  @ApiModelProperty({ type: Object })
  error: any;
}
```

And on my endpoint i have add 
```ts
@ApiResponse({
    status: 200,
    description: '',
    type: ServerApiResponse<PatientCaretrackPopulatedOutput[]>
  })
```

But i have this error `Error TS2348: Value of type 'ServerApiResponse' is not callable.`

So i tried this
```ts
@ApiResponse({
    status: 200,
    description: '',
    type:  new ServerApiResponse<PatientCaretrackPopulatedOutput[]>()
  })
```

But when i launch swagger i have no description of response returned.

It's possible to do this ?

Thx

**Comment by alisherks** [comment]
It's not possible for the same reason as Array<T> or T[] (look at the issue Microsoft/TypeScript/issues/10576).
You could try the following:
```typescript
export abstract class ServiceApiResponse<T> {
  abstract get value(): T;
  
  @ApiModelProperty({ type: Number })
  pages: number;


  @ApiModelProperty({ type: Number })
  page: number;

  @ApiModelProperty({ type: Number })
  maxByPage: number;


  @ApiModelProperty({ type: Number })
  count: number;

  @ApiModelProperty({ type: Object })
  error: any;
}

export class PatientCaretrackPopulatedOutputResponse extends ServiceApiResponse<PatientCaretrackPopulatedResponse[]> {
  @ApiModelProperty({ type: PatientCaretrackPopulatedResponse, isArray: true })
  value: PatientCaretrackPopulatedResponse[];
}
```

**Comment by workfel** [comment]
Hi @alisherks, this works fine.
Thanks

**Comment by omar** [comment]
Cross posting this, does that mean a generic controller wouldn't be able to expose the proper endpoints? For example, something like this (taken from https://github.com/nestjs/nest/issues/228):

``` TypeScript
export abstract class EntityController<T> {
    constructor(protected readonly service: EntityService<T>) { }

    @Get()
    async findAll() {
        return this.service.findAll();
    }

    @Get(':id')
    async getById(@Param() params) {
        return this.service.getById(params.id);
    }

    @Post()
    async create(@Body() data: Partial<T>) {
        return this.service.create(data);
    }
}
```

**Comment by Grymas** [comment]
Running into the same issue, having a generic CRUD controller being inherited is incredibly useful but for PUT and POST methods SwaggerUI is not implicitly picking up the T model.

I've tried specifying the model type with 
`@ImplicitApiBody({ name: 'data', type: thid.entityType })`
where this.entityType is the actual Model declared in the child controller.

I've tried various ways of passing the model constructor to the generic parent controller and have @ImplicitApiBody work, to no avail.

Is there a way for Swagger to find the @Body type "T" in a generic parent controller like CrudController<T> ?
